README.md

```diff
@@ -0,0 +1,434 @@
+
+## queries
+The [**`query`**](#query) function allows you to query your data source and get an object of type [**`CachedQuery`**](#CachedQuery) in return.
+It accepts two parameters, the first is a function with a [**`Fetch`**](#Fetch) signature that is used to retrive data from your data source, the second is an object with the[**`Strategy`**](#Strategy) signature that will be used to decide if the data stored by **Avenger** is still relevant or needs to be refetched.
+
+Although important, `query` is a pretty low level API and **Avenger** offers some convenient utils with a [**`StrategyBuilder`**](#StrategyBuilder) signature that you should prefer over it (unless you have very specific needs):
+
+- **refetch:** runs the fetch function everytime the data is requested.
+- **expire:** when the data is requested, the fetch function is run only if data in the `Cache` is older than the expiration defined, otherwise the cached value is used.
+- **available:** when the data is requested, if a cahed value is available it is always returned, otherwise the fetch function is run and the result stored in the `Cache` accordingly.
+
+All these utils ask you to specify custom `Setoid` instances for the inputs and outputs of the `Fetch` function but you can (and should) use them toghether with one of the built in implementations:
+- **queryShallow:** will use a shallow equality check.
+- **queryStrict:** will use a strict equality check.
+- **queryJSON:** will use a strict equality check after transforming the data via JSON stringification.
+
+Some examples will help clarify:
+```ts
+/*
+  this implementation will always re-run the `Fetch` function
+  even is valid cached data is already present
+  and use shallow equality to compare input output when called
+*/
+const myQuery = queryShallow(fetchFunction, refetch);
+
+/*
+  this implementation will never run the `Fetch` function
+  unless no valid data is present in the Cache
+  and use strict equality to compare input output when called
+*/
+const myQuery = queryStrict(fetchFunction, available);
+
+/*
+  this implementation will run the `Fetch` function only if no valid data is present in the Cache
+  or t > 10000 ms passed till the last time data was fetched
+  and use strict equality to compare input output when called
+*/
+const myQuery = queryJSON(fetchFunction, expire(10000));
+```
+
+Each time the `Fetch` function is run with some `inputs`, those same `inputs` are used as `key` to store the result obtained; from that moment onwards when **Avenger** will need to decide if the data in our [**`Cache`**](#Cache) is present and still valid it will first attempt to retrive it data from the [**`Cache`**](#Cache) by performing an equality check between the `inputs` used to call the `Fetch` function and the keys of the `CacheMap` and then match the result against the cache strategy defined (for instance if we chose `refetch` the data will always be deemed invalid irrespective of the result).
```

Comment:
```suggestion
Each time the `Fetch` function is run with some `inputs`, those same `inputs` are used as `key` to store the result obtained; from that moment onwards when **Avenger** will need to decide if the data in our [**`Cache`**](#Cache) is present and still valid it will first attempt to retrieve its data from the [**`Cache`**](#Cache) by performing an equality check between the `inputs` used to call the `Fetch` function and the keys of the `CacheMap` and then match the result against the cache strategy defined (for instance if we chose `refetch` the data will always be deemed invalid irrespective of the result).
```